### ISSUE TITLE
[Feature Request] Option to disable `F` keybinding (Toggles the Imperium freecam)

### ISSUE THREAD

**Issue by MLIOSBond** [giosuel/imperium#16]
As noted in #1, this mod is impossible to use if you're an `ESDF` player (instead of the classic `WASD`), because the hardcoded into this mod `F` keybinging conflicts with the most basic move: `F` &mdash; which is "strafe right". But this mod shows you the _Imperium freecam_ menu each time you want to strafe right (`F` key).

The basic keybindings for an `ESDF` player are the following:
- `E` &mdash; move forward;
- `S` &mdash; strafe left;
- `D` &mdash; move backward;
- `F` &mdash; strafe right.

So for a quick-fix I would like to suggest an option that disables the `F` keybinding for this mod (so it disables _Imperium freecam_ &mdash; I'm fine with that. Otherwise I just can't use this mod. Or maybe simply introduce a config file (`ini` or something like that) with the most basic mod keybindings: `F1` to `F6`, `T`, `F`, `X`, `Z`).

Thanks for your work!

**Comment by MLIOSBond** [comment]
✔️  Okay, I found a solution.

---

<details>

<summary>

### You'll need

</summary>

- A HEX editor ([wxMEdit][1], for example (here is a [github page][2] for this open-source text/hex editor).
Download it (on Windows you'll need `wxmedit-3.2-win64-bin.zip`) and extract it somewhere on the computer (Desktop is fine. You'll probably use this editor somewhere in the future for other stuff).

</details>

---

<details>

<summary>

### Solution

</summary>


0. Close the game.
    Otherwise the game that is running won't allow us to edit this `Imperium` plugin, because we can't edit things that are currently being used.

1. Find the `dll` file of the current (`Imperium`) plugin.
    If you're using the `Thunderstore Mod Manager`, then you can find the location of the `Imperium` plugin under the `Settings` category in your profile:
    - go to your profile (where you added the `Imperium` plugin);
    - click `Settings`;
    - and then choose `Browse profile folder` &mdash; an explorer should appear;
    - In this folder go to `BepInEx`;
    - then `plugins`;
    - then `giosuel-Imperium`;
    - if you've done everything right, you'll see the `giosuel.Imperium.dll` file. It's the `Imperium` plugin.

    <details>

    <summary>(Screenshot) Thunderstore Mod Manager.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/a28d93cd-c92d-4933-8ea0-0142e4595bef)

    </details>

    The full path of the plugin should look something like this:

    ```
    C:/Users/<username>/AppData/Roaming/Thunderstore Mod Manager/DataFolder/LethalCompany/profiles/<profile-name>/BepInEx/plugins/giosuel-Imperium/giosuel.Imperium.dll
    ```
    where
    - `<username>` &mdash; name of your Windows profile;
    - `<profile-name>` &mdash; is the name of the profile that has the `Imperium` plugin installed.

    <details>

    <summary>(Screenshot) Path to the "giosuel.Imperium.dll" file.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/ae30ce27-15fa-4364-8b37-15bfeae632f5)

    </details>

    Note that you might find `giosuel.Imperium.dll` under the `cache` directory (`Thunderstore Mod Manager/DataFolder/LethalCompany/cache/giosuel-Imperium/0.1.5/giosuel.Imperium.dll`), but you actually need one that's located under the `BepInEx` directory (`Thunderstore Mod Manager/DataFolder/LethalCompany/profiles/Imperium/BepInEx/plugins/giosuel-Imperium`). Please pay attention to that!

2. Open it with your hex editor in the hex mode.

    If you're using `wxMEdit`, then:
    - open it (double click on the `wxmedit.exe` file);
    - click `File`;
    - `Open File...`;
    - choose `giosuel.Imperium.dll` that is located under the directory, that we've found in the step `1`;

    <details>

    <summary>(Screenshot) Open the "giosuel.Imperium.dll`" plugin in the hex editor.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/cd16a603-3c96-491e-ad8f-67c630dba550)

    </details>

    Or simply Drag and Drop the `giosuel.Imperium.dll` file into the `wxMEdit` editor.

    Then after we've successfully opened the plugin, let's switch to the `HEX` view:
    - click `View` on the top;
    - then enable `Hex Mode`.

    Or simply press `Alt-3` to toggle the `HEX` view.

    <details>

    <summary>(Screenshot) Enable the "HEX" view.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/d13b42eb-e65d-4742-bea5-676638dc5312)

    </details>

3. Press `CTRL+F` (or `Search` -> `Find...` in the menu) and check the `Find Hex String` option.

    <details>

    <summary>(Screenshot) Open the "Search" dialog.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/33437a1f-e813-4025-a664-8d8346c1bb3f)

    </details>

4. Put `3C 00 4B 00 65 00 79 00 62 00 6F 00 61 00 72 00 64 00 3E 00 2F 00 66` in this `Search` dialog (if you're curios, it's a [`<Keyboard>/f`][3] string, but in a hexidemical representation).

    <details>

    <summary>(Screenshot) Put the hexademical string into the "Search" dialog.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/40302c66-0780-49f8-8732-94976ac16217)

    </details>

    And click `Find Next`.

    <details>

    <summary>(Screenshot) Find the string.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/a21b02fb-9594-46d9-8a86-5de8b8f7d928)

    </details>

5. Replace the `f` letter that you can see on the right pane (`< K e y b o a r d > / f`) with any key you wish (just click on `f` and press any key).

    <details>

    <summary>(Screenshot) Replace the "f" letter (before).</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/bc11ef8e-5630-459d-a205-c0e51bc1060c)

    </details>

    <details>

    <summary>(Screenshot) Replace the "f" letter (after): I've chosen "p".</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/182af4cf-00b1-4f09-be1c-326e4a84e4dc)

    </details>

    You can also replace other keybindings by searching `3C 00 4B 00 65 00 79 00 62 00 6F 00 61 00 72 00 64 00 3E 00 2F 00` and doing the similar step, that is described above (but instead of the `f` letter there will be, for example, `z` (`K e y b o a r d / z`), `F1` (`K e y b o a r d / F 1`) or `↑` (`K e y b o a r d / u p A r r o w`). You can see all the keybindings in the [`src/Core/ImpInputBindings.cs`][4] file.

    <details>

    <summary>(Screenshot) Other keybindings are there as well. You can edit them too.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/6e8339b1-2bde-4869-bece-72da4158cf9f)

    </details>

6. Save the edited file.

    Simply press `CTRL+S`.

    <details>

    <summary>(Screenshot) Save the file.</summary>

    ![image](https://github.com/giosuel/imperium/assets/29093336/bbbcbd75-7b36-4b36-9098-13a1f34ece49)

    </details>

</details>

---

<details>

<summary>

### Troubleshooting

</summary>

- It should be noted that there might be more than one `Imperium` plugin installed on your computer, because some mods copy files in their directories. So to be 100% sure, just edit one `dll` file and then replace other `Imperium` dll files with the edited one.

</details>

---

<details>

<summary>

### Limitations

</summary>

- You can only replace an old key with a new one if it has an equal or less character count. For example: you can freely replace the `f` key to `p`, but you can't replace `f` with `F7`, because the new key `F7` has more characters in it (`2`), than the previous one (`f`, which is `1`). Another example: you can't replace `upArrow` (7 characters) with `downArrow` (9 characters), but you can freely replace `rightArrow` (10 characters) with `leftArrow` (9 characters). A simple explanation of this limitation is: we're editing a file inplace, so we can't add more bytes/memory to it, we're only allowed to leave the byte count the same (or less, by replacing the rest with zero (`00`) bytes).

</details>

---

<details>

<summary>

### TL;DR

</summary>

What this solution is all about? If we pretend, that a `dll` file is just a config file with extra steps &mdash; then we can manually rebind some keys by applying a solution described above.

</details>


[1]: https://wxmedit.github.io/downloads.html
[2]: https://github.com/wxMEdit/wxMEdit
[3]: https://github.com/giosuel/imperium/blob/ac74daac26e4953f7297a54bb274ecd2b6fc134f/Imperium/src/Core/ImpInputBindings.cs#L20
[4]: https://github.com/giosuel/imperium/blob/ac74daac26e4953f7297a54bb274ecd2b6fc134f/Imperium/src/Core/ImpInputBindings.cs#L17-L44

**Comment by giosuel** [comment]
Thanks for this detailed report on how to fix this issue! I added InputUtils in the new version, meaning you can now rebind all the keybinds to your liking (8b53880098f6abcb4ae4df9e2838a9b2bbc634d8).
